pkg/sfu/buffer/dependencydescriptorparser.go

```diff
@@ -12,91 +13,143 @@ import (
 	onMaxLayerChanged func(int32, int32)
-	decodeTargetLayer []VideoLayer
+	decodeTargets     []DependencyDescriptorDecodeTarget
 }
 
-func NewDependencyDescriptorParser(ddExt uint8, logger logger.Logger, onMaxLayerChanged func(int32, int32)) *DependencyDescriptorParser {
-	logger.Infow("creating dependency descriptor parser", "ddExt", ddExt)
+func NewDependencyDescriptorParser(ddExtID uint8, logger logger.Logger, onMaxLayerChanged func(int32, int32)) *DependencyDescriptorParser {
+	logger.Infow("creating dependency descriptor parser", "ddExtID", ddExtID)
 	return &DependencyDescriptorParser{
-		ddExt:             ddExt,
+		ddExtID:           ddExtID,
 		logger:            logger,
 		onMaxLayerChanged: onMaxLayerChanged,
 	}
 }
 
-func (r *DependencyDescriptorParser) Parse(pkt *rtp.Packet) (*dd.DependencyDescriptor, VideoLayer, error) {
+type DependencyDescriptorWithDecodeTarget struct {
+	Descriptor    *dd.DependencyDescriptor
+	DecodeTargets []DependencyDescriptorDecodeTarget
+}
+
+func (r *DependencyDescriptorParser) Parse(pkt *rtp.Packet) (*DependencyDescriptorWithDecodeTarget, VideoLayer, error) {
+	// DD-TODO: make sure out-of-order RTP packets do not update decode targets
 	var videoLayer VideoLayer
-	if ddBuf := pkt.GetExtension(r.ddExt); ddBuf != nil {
-		var ddVal dd.DependencyDescriptor
-		ext := &dd.DependencyDescriptorExtension{
-			Descriptor: &ddVal,
-			Structure:  r.structure,
-		}
-		_, err := ext.Unmarshal(ddBuf)
-		if err != nil {
-			// r.logger.Debugw("failed to parse generic dependency descriptor", "err", err, "payload", pkt.PayloadType, "ddbufLen", len(ddBuf))
-			return nil, videoLayer, err
-		}
+	ddBuf := pkt.GetExtension(r.ddExtID)
+	if ddBuf == nil {
+		return nil, videoLayer, nil
+	}
 
-		if ddVal.FrameDependencies != nil {
-			videoLayer.Spatial, videoLayer.Temporal = int32(ddVal.FrameDependencies.SpatialId), int32(ddVal.FrameDependencies.TemporalId)
-		}
-		if ddVal.AttachedStructure != nil && !ddVal.FirstPacketInFrame {
-			// r.logger.Debugw("ignoring non-first packet in frame with attached structure")
-			return nil, videoLayer, nil
+	var ddVal dd.DependencyDescriptor
+	ext := &dd.DependencyDescriptorExtension{
+		Descriptor: &ddVal,
+		Structure:  r.structure,
+	}
+	_, err := ext.Unmarshal(ddBuf)
+	if err != nil {
+		// r.logger.Debugw("failed to parse generic dependency descriptor", "err", err, "payload", pkt.PayloadType, "ddbufLen", len(ddBuf))
+		return nil, videoLayer, err
+	}
+
+	if ddVal.FrameDependencies != nil {
+		videoLayer.Spatial, videoLayer.Temporal = int32(ddVal.FrameDependencies.SpatialId), int32(ddVal.FrameDependencies.TemporalId)
+	}
+	if ddVal.AttachedStructure != nil && !ddVal.FirstPacketInFrame {
+		// r.logger.Debugw("ignoring non-first packet in frame with attached structure")
+		return nil, videoLayer, nil
+	}
+
+	if ddVal.AttachedStructure != nil {
+		r.structure = ddVal.AttachedStructure
+		r.decodeTargets = ProcessFrameDependencyStructure(ddVal.AttachedStructure)
```

Comment:
No functional change in this file as well. Just removing several levels of nesting if so that the file is easier to read. Also, making this function to calculate decode targets. I made this as I was initially using it from down track side also for the same functionality. But, once the decode targets are included in `ExtPacket`, this is not called from down track any more, but still leaving it in there as it a clean stand alone function.